lib/copydir.c

```diff
@@ -325,39 +318,31 @@ static int copy_tree_impl (const struct path_info *src, const struct path_info *
 			continue;
 		}
 
-		src_len = strlen (ent->d_name) + 2;
-		dst_len = strlen (ent->d_name) + 2;
-		src_len += strlen (src->full_path);
-		dst_len += strlen (dst->full_path);
-
-		src_name = MALLOC(src_len, char);
-		dst_name = MALLOC(dst_len, char);
-
-		if ((NULL == src_name) || (NULL == dst_name)) {
+		if (asprintf(&src_name, "%s/%s", src->full_path, ent->d_name) == -1)
+		{
 			err = -1;
-		} else {
-			/*
-			 * Build the filename for both the source and
-			 * the destination files.
-			 */
-			struct path_info src_entry, dst_entry;
+			continue;
+		}
+		if (asprintf(&dst_name, "%s/%s", dst->full_path, ent->d_name) == -1)
+		{
+			err = -1;
+			goto skip;
+		}
 
-			(void) snprintf (src_name, src_len, "%s/%s", src->full_path, ent->d_name);
-			(void) snprintf (dst_name, dst_len, "%s/%s", dst->full_path, ent->d_name);
+		src_entry.full_path = src_name;
+		src_entry.dirfd = dirfd(dir);
+		src_entry.name = ent->d_name;
 
-			src_entry.full_path = src_name;
-			src_entry.dirfd = dirfd(dir);
-			src_entry.name = ent->d_name;
+		dst_entry.full_path = dst_name;
+		dst_entry.dirfd = dst_fd;
+		dst_entry.name = ent->d_name;
 
-			dst_entry.full_path = dst_name;
-			dst_entry.dirfd = dst_fd;
-			dst_entry.name = ent->d_name;
+		err = copy_entry(&src_entry, &dst_entry, reset_selinux,
+				 old_uid, new_uid, old_gid, new_gid);
 
-			err = copy_entry(&src_entry, &dst_entry, reset_selinux,
-			                 old_uid, new_uid, old_gid, new_gid);
-		}
-		free (src_name);
-		free (dst_name);
+		free(dst_name);
```

Comment:
Is it strictly necessary? If so, you should have done it in a separate patch, so that it's easier to review the code.